packages/@sanity/form-builder/src/inputs/Reference/ReferenceInput.js

```diff
@@ -203,6 +203,16 @@ export default class ReferenceInput extends React.Component<Props, State> {
     const hasRef = value && value._ref
     const hasWeakMismatch = hasRef && !isMissing && weakIs !== weakShouldBe
 
+    let inputValue
```

Comment:
For readability, maybe you could create a function that takes the preview snapshot and returns the input value?

```js
function resolveInputValue(snapshot) {
  if (snapshot === MISSING_SNAPSHOT) {
    return '<Unpublished or missing document>'
  }
  return snapshot.title || 'Untitled document'
}

// and then when passing the prop:
inputValue={resolveInputValue(previewSnapshot)}
```

Reply:
Tried it, but it always returns 'untitled document' when the component should show the placeholder.

Reply:
Weird. It should be equivalent, except that it should probably be a guard against `snapshot` being `null`. Let me fix it.

Reply:
Actually, the else-if block here will never evaluate as the first if block evaluates to true if `previewSnapshot` is `MISSING_SNAPSHOT`. Something was wonky with the indentation here too. Will fix in `next`

Reply:
You were right, ofc, cause my suggestion did not consider if the input _had a value_.

Not entirely certain it became more readable, but at least it's fixed in `next` (5f2a79b) (squashed my patch into yours in a rebase, together with a fixup of the commit message)